src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs

```diff
@@ -48,21 +78,36 @@ public static bool EncodeHeader(string name, string value, Span<byte> buffer, ou
                 return false;
             }
 
-            if (!EncodeString(value, buffer.Slice(i), out int valueLength, lowercase: false))
+            if (!EncodeString(value, values, separator, buffer.Slice(i), out int valueLength, lowercase: false))
             {
                 return false;
             }
-
             i += valueLength;
 
             length = i;
             return true;
         }
 
-        private static bool EncodeString(string s, Span<byte> buffer, out int length, bool lowercase)
+        private static bool EncodeStringPart(string s, Span<byte> buffer, ref int i, bool lowercase)
         {
             const int toLowerMask = 0x20;
+            // TODO: use huffman encoding
+            for (int j = 0; j < s.Length; j++)
+            {
+                if (i >= buffer.Length)
+                {
+                    return false;
+                }
```

Comment:
It'd be better to do this check once outside of the loop, e.g. create a span from the string, check whether that span is longer than buffer.Length - i, and if it is, slice the span, then iterate over the span.  That way we don't do this check as part of each iteration.